posthog/hogql_queries/actors_query_runner.py

```diff
@@ -212,6 +211,11 @@ def to_query(self) -> ast.SelectQuery:
             else:
                 order_by = []
 
+        # We need to group by all columns in the order by clause
+        for order_expr in order_by:
+            if order_expr.expr not in group_by:
+                group_by.append(order_expr.expr)
```

Comment:
Using `order_by` as a basis of  `group_by` doesn't seem quite right. E.g. in the default People list, we end up grouping by `created_at` this way, which is a waste of resources – and in other cases there may be actual data implications.
We must do `GROUP BY id` in a subquery that's _an input_ for the final aggregations like `count()`, because currently if you add a `count()` column having searched for a user who shows up multiple times, that count ALSO is inflated!

Reply:
Distinct IDs don't have anything to do with this problem of persons being considered multiple times, an `INNER JOIN` cannot _increase_ the number of rows returned. The duplications come from multiple versions of a person being present in the `persons` table.
So, what we need to do is turn that present structure into:
```sql
SELECT <fields - especially aggregations need to be post-grouping, so that each person is only counted once>
FROM
    (
        SELECT id, argMax(properties) AS properties, argMax(<etc>) AS <etc>
        FROM <person table>
        WHERE <team_id plus search conditions - we can do this pre-grouping>
        GROUP BY id
    )
    INNER JOIN
        <person_distinct_id2 table>  -- only when applying search by pdi
    INNER JOIN
        <source query> -- e.g. retention actors
WHERE <property filters, etc - these need to be post-grouping for accuracy>
ORDER BY <...>
```

Reply:
(Unless I misunderstood what you mean by "blowing up", because there definitely is a problem with distinct IDs join blowing queries up performance-wise – less burning than bad data though)

Reply:
Yes, I was being dumb, of course an inner join can increase the number of rows 😄